src/Sylius/Bundle/CoreBundle/OrderProcessing/OrderTaxesApplicator.php

```diff
@@ -0,0 +1,176 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Bundle\CoreBundle\OrderProcessing;
+
+use Sylius\Bundle\CoreBundle\Distributor\IntegerDistributorInterface;
+use Sylius\Bundle\CoreBundle\Provider\DefaultTaxZoneProviderInterface;
+use Sylius\Component\Addressing\Matcher\ZoneMatcherInterface;
+use Sylius\Component\Addressing\Model\AddressInterface;
+use Sylius\Component\Addressing\Model\ZoneInterface;
+use Sylius\Component\Core\Model\AdjustmentInterface;
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Core\Model\OrderItemUnitInterface;
+use Sylius\Component\Resource\Factory\FactoryInterface;
+use Sylius\Component\Taxation\Calculator\CalculatorInterface;
+use Sylius\Component\Taxation\Resolver\TaxRateResolverInterface;
+
+/**
+ * @author Paweł Jędrzejewski <pawel@sylius.org>
+ * @author Mateusz Zalewski <mateusz.zalewski@lakion.com>
+ */
+class OrderTaxesApplicator implements OrderTaxesApplicatorInterface
+{
+    /**
+     * @var CalculatorInterface
+     */
+    protected $taxCalculator;
+
+    /**
+     * @var DefaultTaxZoneProviderInterface
+     */
+    protected $defaultTaxZoneProvider;
+
+    /**
+     * @var FactoryInterface
+     */
+    protected $adjustmentFactory;
+
+    /**
+     * @var IntegerDistributorInterface
+     */
+    protected $integerDistributor;
+
+    /**
+     * @var TaxRateResolverInterface
+     */
+    protected $taxRateResolver;
+
+    /**
+     * @var ZoneMatcherInterface
+     */
+    protected $zoneMatcher;
+
+    /**
+     * @param CalculatorInterface $taxCalculator
+     * @param DefaultTaxZoneProviderInterface $defaultTaxZoneProvider
+     * @param FactoryInterface $adjustmentFactory
+     * @param IntegerDistributorInterface $integerDistributor
+     * @param TaxRateResolverInterface $taxRateResolver
+     * @param ZoneMatcherInterface $zoneMatcher
+     */
+    public function __construct(
+        CalculatorInterface $taxCalculator,
+        DefaultTaxZoneProviderInterface $defaultTaxZoneProvider,
+        FactoryInterface $adjustmentFactory,
+        IntegerDistributorInterface $integerDistributor,
+        TaxRateResolverInterface $taxRateResolver,
+        ZoneMatcherInterface $zoneMatcher
+    ) {
+        $this->taxCalculator = $taxCalculator;
+        $this->defaultTaxZoneProvider = $defaultTaxZoneProvider;
+        $this->adjustmentFactory = $adjustmentFactory;
+        $this->integerDistributor = $integerDistributor;
+        $this->taxRateResolver = $taxRateResolver;
+        $this->zoneMatcher = $zoneMatcher;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function apply(OrderInterface $order)
+    {
+        $this->clearTaxes($order);
+        if ($order->isEmpty()) {
+            return;
+        }
+
+        $zone = $this->provideTaxZone($order->getShippingAddress());
```

Comment:
``getTaxZone``.

In next iterations, I think we should have another interface ``OrderTaxZoneResolver|Provider::getTaxZone($order)``, which would be a nice abstraction layer.

* It would use DefaultTaxZoneProviderInterface + ZoneMatcher and maybe in future it would also allow to use billing address instead of shipping as an option.